File: LC75/1071. Greatest Common Divisor of Strings/GCD.ts
function gcdOfStrings(str1: string, str2: string): string {
    function gcd(a: number, b: number): number {
        if (b === 0) return a;
        return gcd(b, a % b);
    }
    function divisible(s: string, t: string): boolean {
        if (s.length % t.length !== 0) return false;
        const repeatTimes = s.length / t.length;
        let concatenated = '';
        for (let i = 0; i < repeatTimes; i++) {
            concatenated += t;
        }
        return concatenated === s;
    }

    const gcdLength = gcd(str1.length, str2.length);
    const GCDString = str1.substring(0, gcdLength);

    if (divisible(str1, GCDString) && divisible(str2, GCDString)) {
        return GCDString;
    } else {
        return "";
    }
}

File: LC75/1071. Greatest Common Divisor of Strings/GCD.java
public class GCD {
    public String gcdOfStrings(String str1, String str2) {
        if (!(str1 + str2).equals(str2 + str1)) {
            return "";
        }
        int gcdLength = gcd(str1.length(), str2.length());
        return str1.substring(0, gcdLength);
    }

    private int gcd(int a, int b) {
        if (b == 0) {
            return a;
        }
        return gcd(b, a % b);
    }

    public static void main(String[] args) {
        GCD sol = new GCD();
        System.out.println(sol.gcdOfStrings("ABCABC", "ABC")); // Output: "ABC"
        System.out.println(sol.gcdOfStrings("ABABAB", "ABAB")); // Output: "AB"
        System.out.println(sol.gcdOfStrings("LEET", "CODE"));   // Output: ""
    }
}
